# Machine Learning Model Comparison

## Objective

The objective of this notebook is to compare the performance of the machine learning models developed throughout this project for predicting loan performance.

The comparison focuses on:

- Linear Regression
- Standard Logistic Regression
- Balanced Logistic Regression
- Standard Random Forest
- Balanced Random Forest

Each model is evaluated using appropriate performance metrics, with particular emphasis on its ability to identify non-performing loans in a highly imbalanced lending dataset.

The goal is to determine which model provides the most practical solution for supporting credit risk assessment and lending decisions.

In [1]:
# IMPORT LIBRARIES


import pandas as pd

In [2]:
# COMPARISON TABLE


import pandas as pd

comparison = pd.DataFrame({

    "Model":[
        "Linear Regression",
        "Standard Logistic Regression",
        "Balanced Logistic Regression",
        "Standard Random Forest",
        "Balanced Random Forest"
    ],

    "Primary Task":[
        "Regression",
        "Classification",
        "Classification",
        "Classification",
        "Classification"
    ],

    "Accuracy":[
        "-",
        0.982,
        0.697,
        0.982,
        0.982
    ],

    "ROC AUC":[
        "-",
        0.752,
        0.762,
        0.681,
        0.689
    ],

    "Non-Performing Loans Detected":[
        "-",
        "0 / 36",
        "25 / 36",
        "0 / 36",
        "0 / 36"
    ],

    "Recommended for Credit Risk":[
        "No",
        "No",
        "Yes",
        "No",
        "No"
    ]

})

comparison

,Model,Primary Task,Accuracy,ROC AUC,Non-Performing Loans Detected,Recommended for Credit Risk
0,Linear Regression,Regression,-,-,-,No
1,Standard Logistic Regression,Classification,0.982,0.752,0 / 36,No
2,Balanced Logistic Regression,Classification,0.697,0.762,25 / 36,Yes
3,Standard Random Forest,Classification,0.982,0.681,0 / 36,No
4,Balanced Random Forest,Classification,0.982,0.689,0 / 36,No


# Summary of Findings

Five predictive models were developed and evaluated throughout this project to assess their suitability for loan performance prediction and credit risk assessment.

## Linear Regression

The Linear Regression model was developed to predict a continuous lending variable and therefore served a different purpose from the classification models. It demonstrated regression modeling techniques but was not designed to predict loan default.

---

## Standard Logistic Regression

The Standard Logistic Regression model achieved an overall accuracy of **98.2%**, a Precision of **98.2%**, a Recall of **100%**, an F1 Score of **99.1%**, and a ROC AUC score of **0.752**.

Despite these strong performance metrics, the confusion matrix showed that the model classified every loan as **Performing**, failing to identify any of the **36 non-performing loans**.

This demonstrates that the model was heavily influenced by the majority class, making the high accuracy misleading for credit risk assessment.

---

## Balanced Logistic Regression

To address the severe class imbalance, a Balanced Logistic Regression model was developed using `class_weight="balanced"`.

The model achieved:

- Accuracy: **69.7%**
- Precision: **99.2%**
- Recall: **69.7%**
- F1 Score: **81.8%**
- ROC AUC: **0.762**

Unlike the standard model, it successfully identified **25 of the 36 non-performing loans**, representing approximately **69%** of the minority class.

Although this improvement resulted in more false positive predictions and lower overall accuracy, the model became significantly more effective for identifying high-risk borrowers.

---

## Standard Random Forest

The Standard Random Forest classifier achieved:

- Accuracy: **98.2%**
- Precision: **98.2%**
- Recall: **100%**
- F1 Score: **99.1%**
- ROC AUC: **0.681**

However, the confusion matrix revealed that the model classified every loan as **Performing**, failing to detect any of the **36 non-performing loans**.

These results indicate that the model was dominated by the majority class and therefore provided limited practical value for credit risk prediction despite its high overall accuracy.

---

## Balanced Random Forest

The Balanced Random Forest model incorporated `class_weight="balanced"` to improve minority class detection.

The model achieved:

- Accuracy: **98.2%**
- Precision: **98.2%**
- Recall: **100%**
- F1 Score: **99.1%**
- ROC AUC: **0.689**

Although the ROC AUC score increased slightly compared with the Standard Random Forest model, the confusion matrix remained unchanged, with the model continuing to classify every loan as **Performing**.

As a result, the Balanced Random Forest model failed to identify any non-performing loans, indicating that class weighting alone was insufficient to improve the model's performance on this highly imbalanced dataset.

# Final Recommendation

The comparison of the machine learning models demonstrates that **overall accuracy alone is not a reliable measure of model performance** when working with highly imbalanced lending datasets.

Both the Standard Logistic Regression and the Random Forest models achieved very high accuracy (98.2%), yet failed to identify any non-performing loans. Consequently, these models would not be suitable for supporting credit risk assessment despite their impressive accuracy scores.

The Balanced Logistic Regression model produced a lower overall accuracy (69.7%) but successfully identified approximately **69% of the non-performing loans**, making it the most effective model for detecting high-risk borrowers.

Although the Balanced Random Forest model incorporated class weighting, it showed no meaningful improvement over the Standard Random Forest model and continued to classify every loan as performing.

Based on these findings, the **Balanced Logistic Regression model is recommended as the preferred classification model** for this project because it best aligns with the business objective of identifying borrowers with a higher likelihood of default.

Future improvements could include:

- Applying Synthetic Minority Oversampling Technique (SMOTE)
- Hyperparameter tuning
- Threshold optimization
- Cost-sensitive learning
- Advanced ensemble methods such as XGBoost or LightGBM

These techniques may further improve the detection of non-performing loans while reducing false positive predictions.